src/strategies/Staking.sol

```diff
@@ -4,174 +4,348 @@ pragma solidity =0.8.16;
-        debtToken = ERC20(AAVE.getReserveData(address(WETH)).variableDebtTokenAddress);
+    constructor(AffineVault _vault, address[] memory strategists, IBalancerVault _balancer, uint _leverage) AccessStrategy(_vault, strategists) {
+        balancer = _balancer;
+        leverage = _leverage;
 
-        WETH.safeApprove(address(ROUTER), type(uint256).max);
-        WSTETH.safeApprove(address(ROUTER), type(uint256).max);
+        // Dai minting
+        // Send wsteth to join contract
+        ERC20(address(LIDO)).safeApprove(address(WSTETH_JOIN), type(uint256).max);
+       
 
-        WSTETH.safeApprove(address(AAVE), type(uint256).max);
-        aToken.safeApprove(address(AAVE), type(uint256).max);
-        WETH.safeApprove(address(AAVE), type(uint256).max);
-    }
+        // Dai burning / debt payment
+        // Allow Dai join to take Dai from this contract
+        DAI.approve(address(joinDai), type(uint256).max);
 
-    uint256 public lastPosSize;
-    int256 public lastPosPnl;
+        // Compound deposits/borrow
+        DAI.safeApprove(address(cDAI), type(uint256).max);
+        address[] memory cTokens = new address[](1);
+        cTokens[0] = address(cDAI);
 
-    /// @dev Always begin with raw ETH, and 0 WETH
-    function startPosition(uint256 size, uint256 slippageBps, uint256 leverage) external {
-        require(msg.sender == owner, "Staking: only owner can call");
-        // Make sure balance is at least 1.005 * the intended deposit size
-        uint256 ethNeeded = size + size.mulDivUp(slippageBps * 10, 10_000);
-        require(ethNeeded <= address(this).balance, "Staking: insufficient balance");
-        // E.g. if we want to get leverage on 1 Eth, we need about 1.005 to account slippage in WETH -> wstETH trade
-        ERC20[] memory tokens = new ERC20[](1);
-        tokens[0] = WETH;
-        uint256[] memory amounts = new uint256[](1);
-        amounts[0] = size * leverage;
-        IWETH(address(WETH)).deposit{value: ethNeeded}();
-        lastPosSize = ethNeeded;
+        uint256[] memory results = IComptroller(0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B).enterMarkets(cTokens);
+        require(results[0] == 0, "Staking: failed to enter market");
 
-        vault.flashLoan(IFlashLoanRecipient(address(this)), tokens, amounts, abi.encode(Loan.open, size));
-    }
+        // Trade stEth for Eth
+        STETH.safeApprove(address(CURVE), type(uint).max);
 
-    enum Loan {
-        open,
-        close
     }
 
-    function endPosition() external {
-        require(msg.sender == owner, "Staking: only owner can call");
-        // Take loan for amount of debt from aave
+    /// @notice Start a position with `size` principal, taking up to`maxPrincipal` from the user. 
+    function startPosition(uint256 size, uint256 maxPrincipal) external onlyRole(STRATEGIST_ROLE) {
+        require(maxPrincipal >= size, "Staking: max > size");
+        WETH.safeTransferFrom(msg.sender, address(this), maxPrincipal);
```

Comment:
Should we pulling the asset from vault or strategist?

Reply:
Deleted this, it was just for some easy testing.